File: LargestValueEachTreeRow.java
class LargestValueEachTreeRow{

    // BFS
    public List<Integer> largestValues(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if(root == null) return result;
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);

        while(!q.isEmpty()){
            int size = q.size();
            int max = Integer.MIN_VALUE;
            for(int i=0; i<size; i++){
                TreeNode cur = q.poll();
                max = Math.max(max, cur.val);
                if(cur.left != null) q.add(cur.left);
                if(cur.right != null) q.add(cur.right);
            }
            result.add(max);
        }
        return result;
    }

    // DFS
    List<Integer> result;
    public List<Integer> largestValues(TreeNode root) {
        result = new ArrayList<>();
        if(root == null) return result;
        dfs(root, 0);
        return result;
    }

    private void dfs(TreeNode root, int level){
        // base
        if(root == null) return;


        // logic
        if(result.size() == level){
            result.add(root.val);
        } else{
            int prev = result.get(level);
            int max = Math.max(prev, root.val);
            result.set(level, max);
        }
        dfs(root.left, level +1);
        dfs(root.right, level +1);
    }
}

File: Largest_Value_Each_Tree_Row.py
# Time complexity: O(n)
# Space complexity: O(n)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def largestValues(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
		# Base condition
        if not root or root == None:
            return []
       
	   # append root to queue
        queue = deque()
        queue.append(root)
        result = [] 
        
		# Perform till queue not empty
        while queue:
            queue_len = len(queue)
			
			# Take max_val to compare
            max_val = float("-inf")
            for i in range(queue_len):
                node = queue.popleft()
             
				# Get max value by comparing max_val and node.val
                max_val = max(max_val, node.val)
				
				# Load NODE.LEFT to queue
                if node.left != None:
                    queue.append(node.left)
				
				# Load NODE.RIGHT to queue
                if node.right != None:
                    queue.append(node.right)
            
			# Append my max_val to result
            result.append(max_val)
            
            
        return result
